### ISSUE TITLE
Structural Interfaces (not behavioral)

### ISSUE THREAD

**Issue by zot** [JuliaLang/julia#43083]
I think it would be helpful if there were "structural interfaces" to express that a value is expected to have readable or writable fields of certain types when the value could also contain other fields (so the structural interface could be a subset of the value's type). Maybe something like `Readable{x::Int, y::String}` and `Writable{z::Float64}` (those names would probably clash with a lot of code that's already out there). This is a way to sort of embrace Julia's duck typing within the type system (it's no coincidence that TypeScript's interfaces can contain fields). This isn't like Java's interfaces because it doesn't address methods but I think it would provide some important utility.

This would let a function author indicate, "this function will fail if the input does not have an `x::Int` field," and IDEs could warn developers when they try to pass incompatible data to the function.

This could help document data used in protocols, file formats, etc. but still allow developers the freedom to use structures or named tuples which contain more information than the function uses.

**Comment by JeffBezanson** [comment]
Very related: #4935

**Comment by zot** [comment]
True but structs and named tuples can be type compatible with many structural interfaces at once.

**Comment by iago-lito** [comment]
Would this be actually desugared to a behavioural interface with inlined `get_field_x() -> Int` and `set_field_x(Int)` methods?

**Comment by zot** [comment]
I don't think this should simply be sugar, I think it should be type information that can be checked with `isa`, etc. so that this:
```julia
structinterface AStructuralInterface
  a::Number
end
```
Guarantees that `a` exists in anything implementing the interface and
```
function foo(x::AStructuralInterface)
	println(x isa AStructuralInterface)
	println(x.a)
end
```
will work for any `x` that implements AStructuralInterface, printing `true` and the value of `a`.

**Comment by vtjnash** [comment]
Should we convert this to a "Discussion"?

**Comment by zot** [comment]
I certainly think it's worth discussing! :smiley:
